gen/map.go

```diff
@@ -190,6 +191,83 @@ func (m *mapGenerator) Reader(g Generator, spec *compile.MapSpec) (string, error
 	return name, wrapGenerateError(spec.ThriftName(), err)
 }
 
+// Equals generates a function to compare maps of the given type
+//
+// func $name(lhs, rhs $mapType) bool {
+//      ...
+// }
+//
+// And returns its name.
+func (m *mapGenerator) Equals(g Generator, spec *compile.MapSpec) (string, error) {
+	name := "_" + valueName(spec) + "_Equals"
+	if m.HasEquals(name) {
+		return name, nil
+	}
+
+	err := g.DeclareFromTemplate(
+		`
+            <$mapType := typeReference .Spec>
+
+            <$lhs := newVar "lhs">
+            <$rhs := newVar "rhs">
+            <$i := newVar "i">
+            <$j := newVar "j">
+            <$lk := newVar "lk">
+            <$lv := newVar "lv">
+            <$rk := newVar "rk">
+            <$rv := newVar "rv">
+            <$ok := newVar "ok">
+			func <.Name>(<$lhs>, <$rhs> <$mapType>) bool {
+				if len(<$lhs>) != len(<$rhs>) {
+					return false
+				}
+
+				<if isHashable .Spec.KeySpec>
+					for <$lk>, <$lv> := range <$lhs> {
+						<$rv>, <$ok> := <$rhs>[<$lk>]
+						if !<$ok> {
+							return false
+						}
+						if !(<equals .Spec.ValueSpec $lv $rv>) {
```

Comment:
Consider breaking out some of the logic into functions here. For example,

```
<if isHashable .Spec.KeySpec>
  ...
<else>
  <equalsForUnhashableKey ...>
<end>
```

Where you inject `equalsUnhashable` into *just* this template by using TemplateFunc, implementing the function itself as a method on `mapGenerator`

```
g.DeclareFromTemplate(
  ...,
  TemplateFunc("equalsForUnhashableKey", m.equalsForUnhashableKey)
)
```

```
func (m *mapGenerator) equalsForUnhashableKey(g Generator, ...) (string, error) {
  // ...
}
```